### PR TITLE
doc: add documentation for tool targets

### LINKED ISSUE #26758: Missing documentation of report targets (ram/rom report, puncover)
**Describe the bug**
In Zephyr build system there exists additional report targets, ram/rom report, puncover, pahole.

But to new Zephyr users, those targets are very hidden and not documented.
In order to discover those targets, one must look into:
https://github.com/zephyrproject-rtos/zephyr/blob/master/cmake/reports/CMakeLists.txt

We should document such targets, maybe under:
https://docs.zephyrproject.org/latest/guides/coverage.html

**Expected behavior**
A page describing report target in Zephyr documentation.

**Impact**
Features in Zephyr not discovered (and thus probably not used).

This means new users could risk wasting time on re-written same features downstream.

Edit: pahole added to list.

### PR DESCRIPTION
**Pull request by nashif**
Document the following targets:
- ram_report
- rom_report
- puncover
- pahole

Fixes #26758

### CODE REVIEW

File: doc/guides/optimizations/tools.rst [hunk 1]
```diff
@@ -0,0 +1,209 @@
+.. _optimization_tools:
+
+Optimization Tools
+####################
```
Comment (henrikbrixandersen) [inline doc/guides/optimizations/tools.rst:4]: Too many hashmarks here.

File: doc/guides/optimizations/tools.rst [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+
+.. _footprint_tools:
+
+Footprint and Memory Usage
+**************************
+
+The build system offers 3 targets to view and analyse RAM, ROM and Stack usage
```
Comment (henrikbrixandersen) [inline doc/guides/optimizations/tools.rst:11]: Was `Stack` intentionally capitalised here?

File: doc/guides/optimizations/tools.rst [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+in generated images. The tools run on the final image and give and information
```
Comment (henrikbrixandersen) [inline doc/guides/optimizations/tools.rst:12]:
```suggestion
in generated images. The tools run on the final image and give information
```

File: doc/guides/optimizations/tools.rst [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+about size of symbols and code being used in both RAM and ROM. Additionally and
+with features available through the compiler we can also generate worst-case
+stack usage analysis:
```
Comment (henrikbrixandersen) [inline doc/guides/optimizations/tools.rst:15]:
```suggestion
about size of symbols and code being used in both RAM and ROM. Additionally,
with features available through the compiler, we can also generate worst-case
stack usage analysis:
```

File: doc/guides/optimizations/tools.rst [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+
+Tools that are available as build system targets:
+
+Build Target: puncover
+======================
+
+This target uses a 3rd party tools called puncover which can be found `here
+<https://github.com/HBehrens/puncover>`_. When this target is built, it will
+launch a local web server which will allow you to open a web client and browser
```
Comment (henrikbrixandersen) [inline doc/guides/optimizations/tools.rst:24]:
```suggestion
launch a local web server which will allow you to open a web client and browse
```

File: doc/guides/optimizations/tools.rst [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+the files and view their ROM,RAM and Stack usage. Before you can use this
```
Comment (henrikbrixandersen) [inline doc/guides/optimizations/tools.rst:25]:
```suggestion
the files and view their ROM, RAM and stack usage. Before you can use this
```

File: doc/guides/optimizations/tools.rst [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+
+
+.. zephyr-app-commands::
+    :tool: all
+    :app: samples/hello_world
+    :board: reel_board
+    :goals: pahole
+
+
+After running this target, pahole will out the results to the console::
```
Comment (henrikbrixandersen) [inline doc/guides/optimizations/tools.rst:179]:
```suggestion
After running this target, pahole will output the results to the console::
```

File: doc/guides/optimizations/tools.rst [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+Data Structures
+****************
+
+
+Build Target: pahole
+=====================
+
+Poke-a-hole (pahole) is an object-file analysis tool to find the size of
+the data structures, and the holes caused due to aligning the data
+elements to the word-size of the CPU by the compiler.
```
Comment (henrikbrixandersen) [inline doc/guides/optimizations/tools.rst:169]: We should probably mention that `pahole` must be installed prior to using this target - and mention where to obtain it (e.g. `https://git.kernel.org/pub/scm/devel/pahole/pahole.git` and perhaps `sudo apt-get install dwarves`).

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]
Thanks, this is a welcome addition to the documentation. A few nits below:

**Review by henrikbrixandersen — APPROVED** [review]

**Review by dcpleung — APPROVED** [review]

**Pull request merged**